test/fika/erl_translate_test.exs

```diff
@@ -164,6 +164,38 @@ defmodule Fika.ErlTranslateTest do
             ]} = result
   end
 
+  describe "string interpolation" do
+    test "replaces with string" do
+      str = ~S"""
+      "#{"Hello"} #{"World"}"
+      """
+
+      ast = TestParser.expression!(str)
+      result = ErlTranslate.translate_expression(ast)
+
+      assert {
+               :string,
+               1,
+               [
+                 {:string, 1, 'Hello'},
+                 {:string, 1, ' '},
+                 {:string, 1, 'World'}
+               ]
+             } = result
+    end
+
+    test "handles expressions" do
+      str = ~S"""
+      "#{2 + 2}"
+      """
+
+      ast = TestParser.expression!(str)
+      result = ErlTranslate.translate_expression(ast)
+
+      assert {:string, 1, [{:op, 1, :+, {:integer, 1, 2}, {:integer, 1, 2}}]} = result
```

Comment:
This will result in a non-printable string instead of `"4"` because this will ultimately try to parse as `{:string, 1, [4]}`.
It seems to me that an intermediate kernel conversion might me needed for literal expressions like this. Maybe Erlang Abstract Format already has something like this?

Reply:
That's right. My approach would be to 

1. parse "foo #{bar} baz" into
```
{
  :string,
  line,
  [
      "foo"
      {:identifier, line, :bar},
      "baz"
  ]
}
```

2. Make infer_exp in TypeChecker for :string to ensure all non binary items in the list are of type string

3. In erl_translate for :string, translate strings into `{:bin, line, [{:bin_element, line, exp, :default, :default}, ...]}` for translating strings.

To understand how Elixir translates strings, run this in iex:

```elixir
q = """
bar = "bar"
"foo #\{bar} baz"
"""
q = Code.string_to_quoted!(q)
:elixir.quoted_to_erl(q, :elixir.env_for_eval([]))
```

You can ignore the case clauses there. What elixir does there is - if bar is binary, use the variable as is, else use `to_string(bar)` . We don't need this.

Reply:
@emilsoman I've pushed items 1 (modified for multiple exps) & 3, but I'm a little bit confused on how to implement item 2. I've dug `TypeChecker` and found `infer_list_exps`, do we need something like that for strings or it's more simple than that?

Reply:
Something like this:

```elixir
  def infer_exp(env, {:string, _line, string_parts}) do
    Enum.reduce_while(string_parts, env, fn
      string, env when is_binary(string) ->
        Logger.debug("String #{string} found. Type: String")
        {:cont, {:ok, "String", env}}

      exp, env ->
        Logger.debug("String interpolation found. Inferring type of expression")

        case infer_exp(env, exp) do
          {:ok, "String", env} = result ->
            {:cont, result}

          {:ok, other_type, _env} ->
            message =
              "Expression used in string interpolation expected to be String, got #{other_type}"

            {:halt, {:error, message}}

          error ->
            {:halt, error}
        end
    end)
  end
```